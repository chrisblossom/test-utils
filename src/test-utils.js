/* @flow */

/* eslint-disable no-param-reassign */

import {
    cloneDeep,
    isArray,
    forIn,
    isPlainObject,
    isError,
    isObjectLike,
    isString,
} from 'lodash';

// https://stackoverflow.com/a/1527820
function getRandomInt(min: number = 1, max: number = 1000000000) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

type RemoveStringArgs = {
    source: *,
    pattern: string,
    replaceWith?: string,
};

function removeString(args: RemoveStringArgs) {
    const { source, pattern = '', replaceWith = '' } = args;

    if (!isObjectLike(source) && !isString(source)) {
        return source;
    }

    const regex = new RegExp(pattern, 'g');

    /**
     * Unknown how to modify the error.stack. Just return the message for testing purposes with an Error: key
     */
    if (isError(source)) {
        return {
            Error: source.message.replace(regex, replaceWith),
        };
    }

    const newObj = cloneDeep(source);

    /**
     * Handle nested errors
     *
     * https://gist.github.com/tushariscoolster/567c1d22ca8d5498cbc0
     */
    const traverse = (obj) => {
        if (isString(obj)) {
            return obj.replace(regex, replaceWith);
        }

        return forIn(obj, (val, key) => {
            if (isArray(val)) {
                val.forEach((el, index) => {
                    if (isPlainObject(el)) {
                        traverse(el);
                    }

                    if (isError(el)) {
                        obj[key][index] = {
                            Error: el.message.replace(regex, replaceWith),
                        };
                    }

                    if (isString(el)) {
                        obj[key][index] = el.replace(regex, replaceWith);
                    }
                });
            }

            if (isPlainObject(val)) {
                traverse(obj[key]);
            }

            if (isError(val)) {
                obj[key] = {
                    Error: val.message.replace(regex, replaceWith),
                };
            }

            if (isString(val)) {
                obj[key] = val.replace(regex, replaceWith);
            }
        });
    };

    const result = traverse(newObj);

    return result;
}

const normalizeRootPath = (source: any = '', cwd: ?string) => {
    const pattern = cwd || process.cwd();

    return removeString({
        source,
        pattern,
        replaceWith: '/normalized/root/path',
    });
};

function getFunctionNames(mergedPreset: any) {
    const data = mergedPreset;

    const getName = (item) => {
        if (isPlainObject(item) && item.name && item.task) {
            return item.name;
        }

        if (isObjectLike(item)) {
            return item;
        }

        if (isString(item)) {
            return item;
        }

        return item.name;
    };

    const result = Object.keys(data).reduce((acc, current) => {
        const matched = data[current];
        if (isPlainObject(matched)) {
            return { ...acc, [current]: matched };
        }

        const updated = matched.map((fn) => {
            if (Array.isArray(fn)) {
                return fn.map((arrayOfFunctions) => {
                    return getName(arrayOfFunctions);
                });
            }

            return getName(fn);
        });

        return { ...acc, [current]: updated };
    }, {});

    return normalizeRootPath(result);
}

export { getFunctionNames, getRandomInt, normalizeRootPath, removeString };
