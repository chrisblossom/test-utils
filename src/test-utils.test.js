/* @flow */

import path from 'path';
import { normalizeRootPath, removeString } from './test-utils';

describe('removeString', () => {
    it('removes from object', () => {
        const source = {
            one: {
                two: {
                    three: '/remove/this/but/not/this/',
                },
            },
        };

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toMatchSnapshot();
    });

    it('removes from string', () => {
        const source = '/remove/this/but/not/this/';

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toEqual('but/not/this/');
    });

    it('adds replaceWith', () => {
        const source = '/remove/this/but/not/this/';

        const pattern = '/remove/this/';

        const result = removeString({
            source,
            pattern,
            replaceWith: '/replace/with/this/',
        });

        expect(result).toEqual('/replace/with/this/but/not/this/');
    });

    it('handles undefined', () => {
        const source = undefined;

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toEqual(undefined);
    });

    it('handles Error', () => {
        const source = new Error('/remove/this/but/not/this/');

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toMatchSnapshot();
    });

    it('handles Error deep in array', () => {
        const source = [
            '/remove/this/but/not/this/first/',
            new Error('/remove/this/but/not/this/second'),
        ];

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toMatchSnapshot();
    });

    it('handles Errors deep in object', () => {
        const source = {
            inside: {
                another: {
                    error: new Error('/remove/this/but/not/this/'),
                },
                arr: [
                    '/remove/this/but/not/this/first/',
                    new Error('/remove/this/but/not/this/second/'),
                ],
            },
        };

        const pattern = '/remove/this/';

        const result = removeString({ source, pattern });

        expect(result).toMatchSnapshot();
    });
});

describe('normalizeRootPath', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    it('removes rootPath', () => {
        const dir = path.resolve(__dirname, '__sandbox__/');
        process.chdir(dir);

        const source = {
            str: path.resolve(dir, 'file.js'),
        };
        const result = normalizeRootPath(source);

        expect(result).toMatchSnapshot();
    });
});
