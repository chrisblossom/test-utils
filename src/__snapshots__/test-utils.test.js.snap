// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeRootPath removes rootPath 1`] = `
Object {
  "str": "/normalized/root/path/file.js",
}
`;

exports[`removeString handles Error 1`] = `
Object {
  "Error": "but/not/this/",
}
`;

exports[`removeString handles Error deep in array 1`] = `
Array [
  "but/not/this/first/",
  Object {
    "Error": "but/not/this/second",
  },
]
`;

exports[`removeString handles Errors deep in object 1`] = `
Object {
  "inside": Object {
    "another": Object {
      "error": Object {
        "Error": "but/not/this/",
      },
    },
    "arr": Array [
      "but/not/this/first/",
      Object {
        "Error": "but/not/this/second/",
      },
    ],
  },
}
`;

exports[`removeString removes from object 1`] = `
Object {
  "one": Object {
    "two": Object {
      "three": "but/not/this/",
    },
  },
}
`;
